src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1442,7 +1448,14 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, BindingD
                 op1 = InferTypeForDiscardAssignment((BoundDiscardExpression)op1, op2, diagnostics);
             }
 
-            return BindAssignment(node, op1, op2, isRef, diagnostics);
+            BoundAssignmentOperator result = BindAssignment(node, op1, op2, isRef, diagnostics);
+
+            if (forceDynamicResult)
+            {
+                result = result.Update(result.Left, result.Right, result.IsRef, Compilation.DynamicType);
```

Comment:
If we add a common helper for `AdjustAssignmentType`, we could do this update unconditionally (it should be lazy if nothing changed). Something like:
`result = result.Update(..., AdjustAssignmentType(..., forceDynamicResult));`

Reply:
> If we add a common helper for AdjustAssignmentType, we could do this update unconditionally (it should be lazy if nothing changed). 

The `Update` API is node specific and we are adjusting types for different nodes, we cannot share the same `Update` call for all of them. Also, it is more efficient to set the final type on the node right from the start, instead of updating it later. We are doing that most of the time. I will keep this logic as is.